FIRES Integration - Moderation Recommendations Feed and General Retractions (#71)
eigenmagic/fediblockhole: README.md

### State tracking and retractions

FediBlockHole maintains a JSON state file to track its position in each
dataset's changes feed. On the first run, it fetches the full snapshot. On
subsequent runs, it polls only for new changes since the last run.

When a FIRES dataset publishes a **retraction** (meaning "we no longer recommend
blocking this domain"), FediBlockHole removes that domain from the source's
contribution to the merge. If other sources still recommend blocking it, the
block remains. Retractions only affect the source that issued them.

The state file defaults to `~/.fediblockhole/fires_state.json`. You can override
this with the `fires_state_file` config option or the `--fires-state-file`
commandline flag.

### Labels as comments

FIRES recommendations include labels from the
[IFTAS shared vocabulary](https://about.iftas.org/library/shared-vocabulary-labels/)
(e.g., "Hate Speech", "CSAM", "Spam"). These are mapped to the `public_comment`
field on domain blocks, so instance admins can see why a domain was recommended

> for blocking.

Reviewer: This is just one known label vocabulary, based on the Digital Trust & Safety Partnerships' labels which they released as CC-BY license. I know for instance that [garden fence](https://gardenfence.github.io/) has it's own label vocabulary. Others may exist in the future too.

### The `accept` policy

The FIRES protocol includes an `accept` policy for recommending that a domain
*should* be federated with. FediBlockHole handles `accept` the same way it
handles its existing allowlists: domains with an `accept` recommendation are
removed from the merged blocklist before it is pushed to instances.

This means an `accept` from a FIRES dataset acts as an override, the same as
adding a domain to a CSV allowlist. It does not call any instance API to

> explicitly allow the domain — it simply prevents it from being blocked.

Reviewer: This is correct, if you're not doing federation policies and just the binary approach that Mastodon uses of domain blocks OR domain allows.

If you don't want FIRES `accept` policies to influence your blocklist at all,
set `ignore_accept = true` on the source:

```toml
blocklist_fires_sources = [
  { server = 'https://fires.example.com', ignore_accept = true },
]
```

With `ignore_accept` enabled, `accept` recommendations are silently skipped.
Block recommendations (`drop`, `reject`, `filter`) and retractions still work

> normally.

Reviewer: This is incorrect, as it is plausible and possible for a `Recommendation` of `drop` to become a `Recommendation` of `accept` without an intermediary `Retraction`

> ### Retractions: removing blocks that are no longer recommended

Reviewer: ```suggestion
### Retractions: removing data that is no longer recommended or advised
```

Historically, FediBlockHole has been additive — it adds and updates blocks but
never removes them. This is safe but means blocks stay on your instance forever,
even if every source stops recommending them.

FIRES changes this by providing the state that blocklists never had. When a
FIRES dataset retracts a recommendation, FediBlockHole can now act on it.

There are two retraction mechanisms, and they can be used together:

#### Source-level retractions (`retractions = true`)

This is the FIRES-native approach. When a trusted FIRES source explicitly
retracts a domain, the block is removed from your instance — **regardless of
who originally added it** — as long as no other source in your merged list still

> recommends blocking it.

Reviewer: You would almost certainly want to do bookkeeping to check "did this dataset add this domain as a recommendation/advisory?" as to be able to know if the `retraction` is valid. Otherwise a retraction from one dataset may override a recommendation from another.

If there is a case of one dataset has the latest state for an entity as `retraction` or `tombstone` and another dataset has a `advisory` or `recommendation`, you have a few ways of merging that: manual merging, least permissive (i.e., most sever policy applies) or most permissive (least sever policy applies).

You could also do automatic merging and fall back to requesting an operator resolve a merge conflict were an appropriate final policy cannot be determined.
